### DISCUSSION TITLE
Adopt twoslash for docs

### DISCUSSION THREAD

**Discussion by AndreyYolkin** [radashi-org/radashi#298] [Ideas]
As one of the aims of radashi is better Typescript support, we can highlight it better with twoslash integration, available by brand-new community plugin https://github.com/MatthiesenXYZ/EC-Plugins/tree/main/packages/twoslash, which is a [twoslash](https://twoslash.netlify.app/) wrapper for Expressive Code, used in Astro.

I've tried to adopt some examples and how it may look like:

### _.assign
<img width="785" alt="image" src="https://github.com/user-attachments/assets/459559d2-3d01-40b3-9ab5-4dd654317a2b">

### _.castArray
<img width="771" alt="image" src="https://github.com/user-attachments/assets/e0b9500d-465a-46bf-bb61-9ab6c7d938df">

### _.first (we can discuss the proper naming)
<img width="789" alt="image" src="https://github.com/user-attachments/assets/9d0a9fc8-d9cc-4c48-b932-08e5170ec776">

The best part is that types are picked from code automatically and we don't need to specify any additional data (only mocks, like in _.all examples)

<img width="795" alt="image" src="https://github.com/user-attachments/assets/c73d5138-13db-4a5f-bb5c-5335b7464c98">

**Comment by aleclarson** [comment]
Love it. Does twoslash get loaded after first paint?

**Reply by AndreyYolkin** [reply]
It took part in rendering, but extra js logic is included in Expressive Code script, which is loaded as module and not blocking render
<img width="1728" alt="image" src="https://github.com/user-attachments/assets/a70d4fa2-7381-4fa1-9583-81caab4a2e96">

**Reply by AndreyYolkin** [reply]
I think, i can make a custom build and provide a link with some examples with twoslash enabled

**Reply by AndreyYolkin** [reply]
UPD: we are live
Check the examples at https://luminous-elf-a6e4e4.netlify.app/ from alphabetical to fork is ready, more to come if we decide that's useful

**Reply by aleclarson** [reply]
Sweet. I think it's a worthy feature. PR welcome :)

**Reply by Adammatthiesen** [reply]
Hey @AndreyYolkin Thanks for supporting my project (`Expressive-code-twoslash`)!

Currently working on rolling out a nice feature update(bringing more of twoslash into EC) as well as some huge code cleanup, the JS module mentioned is handling hiding and showing the popups using a unique ID system so that the hidden popups don't cause any screen reader issues, as the current goal is to not only bring Twoslash to EC, but to also add accessibility into Twoslash rendering, since this is a custom implementation based on the `Shiki-twoslash` and not a direct wrapper of the shiki version

**Reply by AndreyYolkin** [reply]
Hey, @Adammatthiesen, and thanks for coming, I saw a great opportunity to touch the technologies that were on my radar. 

I see that radashi uses patched version of starlight, imported as git submodule. I think I will open another proposal to utilise pnpm's ability to patch package (with https://github.com/antfu/pnpm-patch-i it has a better DX as well) and maintain only patch instead of almost a fork of starlight repo. Once this is done, it will be easier for the docs to be up to date. 

I suppose that by the time I manage how to do contributions easier for other developers you also made the things you mentioned (I saw that this repo is almost a week old, but found that it's already in great condition).

**Comment by AndreyYolkin** [comment]
I'm making a progress with translation, still there are some questions:

**1.** I needed to tweak .npmrc with `shamefully-hoist=true` to complete installation with pnpm. I will think later about committing lockfile, maybe the instruction in README is not exhaustive enough, as simple copy-pasting didn't work.
**2.** Sometimes like in `defer example` we will need to define fake functions in order to make twoslash show types. I think this is better than just leave the code section without twoslash

### Render:
<img width="400" alt="image" src="https://github.com/user-attachments/assets/867f414f-5ac5-48cb-aeaa-f1c84a7642b7">

### MDX:
<img width="400" alt="image" src="https://github.com/user-attachments/assets/234ea702-3772-4cb3-a85e-eb7be3057042">

**3.** Although there are still some scenarios where it's hard to define even mocks, like code extracted from `lodash` library. I decided to leave this code as-is, but we may reconsider it in a future.
<img width="400" alt="image" src="https://github.com/user-attachments/assets/df5dd499-d224-456b-92cf-11386a4a4915">

**4.** (Opinionated) I propose to add `as const` to data mocks, to illustrate type-narrowing in radashi even better **OR** include both options
### Array.First:
<img width="400" alt="image" src="https://github.com/user-attachments/assets/75e3c66e-5227-430f-b4ed-d95f7d8c399d">

### Array.First(with as const):
<img width="400" alt="image" src="https://github.com/user-attachments/assets/f53033be-ef06-4e67-ada5-03163b1179bc">

**5.** AFAIU there will be two PR's: for website docs and for radashi docs. I can open PR to the website builder now, as it's almost ready (except p.1) and draft PR to the docs examples

**Reply by AndreyYolkin** [reply]
@aleclarson sorry for pinging, I need my thoughts to be validated + decision for p. 4 in order to continue adopting twoslash. I will be grateful if you can review and make a comments about 🙂

**Reply by aleclarson** [reply]
@AndreyYolkin The ping was 100% warranted. Sorry to leave you hanging!

We're agreed on points 2, 3, and 4. :+1:

Probably best to do only `as const`, instead of having examples with and without; just to keep the number of examples at a manageable level. Hopefully the `as const` assertions don't confuse the TypeScript newbies too much.


> AFAIU there will be two PR's: for website docs and for radashi docs. I can open PR to the website builder now, as it's almost ready (except p.1) and draft PR to the docs examples

Makes sense to me! Your work on this is very much appreciated 🙂

> I needed to tweak .npmrc with `shamefully-hoist=true` to complete installation with pnpm. I will think later about committing lockfile, maybe the instruction in README is not exhaustive enough, as simple copy-pasting didn't work.

You're referring to the instructions [here](https://github.com/radashi-org/radashi-org.github.io/blob/a5375235d6cb748ded3052d1155805d6dd76c8c8/README.md?plain=1#L4-L8), right?

I just tested them and got this error when running `pnpm dev`. Everything before that seemed to work. If this can be fixed without `shamefully-hoist`, that would be ideal. But since you have a workaround, we can address the issue later.

```
[vite] Error when evaluating SSR module /starlight/packages/starlight/index.ts: failed to import "@astrojs/mdx"
|- Error: Cannot find module '@astrojs/mdx' imported from '/Users/alec/dev/sandbox/radashi-org.github.io/starlight/packages/starlight/index.ts'
```

**Reply by AndreyYolkin** [reply]
+1 on that error, I followed the steps you provided and got the same error, that's why I ended with hoisting. I'll try to find a way better than this later. Back to the grind!

**Reply by aleclarson** [reply]
I'm looking forward to this improvement. Let me know if you'd like assistance, @AndreyYolkin. 😄

**Reply by AndreyYolkin** [reply]
Sorry for delay, I think I'll open PR until February. Thank you for offering a help 🤍

**Reply by AndreyYolkin** [reply]
@aleclarson, I've updated the mirror at https://luminous-elf-a6e4e4.netlify.app/, so it's nearly complete. Please take a look at it.
In some edge cases like [Async -> All](https://luminous-elf-a6e4e4.netlify.app/reference/async/all/) I silenced the errors, that's why lots of objects has `any`, please notify if you want to have better types here.

**Reply by aleclarson** [reply]
@AndreyYolkin Hey, great to hear back from you! It's looking great. We can save the edge cases for another PR. It's not a big deal, I think. 😄
